homeassistant/components/sun/__init__.py

```diff
@@ -83,57 +102,105 @@ def state_attributes(self):
+            utc_point_in_time, 'dawn', PHASE_NAUTICAL_TWILIGHT)
+        self.next_rising = self._check_event(
+            utc_point_in_time, SUN_EVENT_SUNRISE, PHASE_TWILIGHT)
+        self.location.solar_depression = -10
+        self._check_event(utc_point_in_time, 'dawn', PHASE_SMALL_DAY)
+        self.next_noon = self._check_event(
+            utc_point_in_time, 'solar_noon', None)
+        self._check_event(utc_point_in_time, 'dusk', PHASE_DAY)
+        self.next_setting = self._check_event(
+            utc_point_in_time, SUN_EVENT_SUNSET, PHASE_SMALL_DAY)
+        self.location.solar_depression = 'civil'
+        self.next_dusk = self._check_event(
+            utc_point_in_time, 'dusk', PHASE_TWILIGHT)
+        self.location.solar_depression = 'nautical'
+        self._check_event(
+            utc_point_in_time, 'dusk', PHASE_NAUTICAL_TWILIGHT)
+        self.location.solar_depression = 'astronomical'
+        self._check_event(
+            utc_point_in_time, 'dusk', PHASE_ASTRONOMICAL_TWILIGHT)
+        self.next_midnight = self._check_event(
+            utc_point_in_time, 'solar_midnight', None)
+
+        # Need to calculate phase if next is noon or midnight
+        if self.phase is None:
+            elevation = self.location.solar_elevation(self._next_change)
+            if elevation >= 10:
+                self.phase = PHASE_DAY
+            elif elevation >= 0:
+                self.phase = PHASE_SMALL_DAY
+            elif elevation >= -6:
+                self.phase = PHASE_TWILIGHT
+            elif elevation >= -12:
+                self.phase = PHASE_NAUTICAL_TWILIGHT
+            elif elevation >= -18:
+                self.phase = PHASE_ASTRONOMICAL_TWILIGHT
+            else:
+                self.phase = PHASE_NIGHT
+
+        self.rising = self.next_noon < self.next_midnight
+
+        _LOGGER.info(
+            "sun phase_update@%s: phase=%s",
+            utc_point_in_time.isoformat(),
+            self.phase,
+        )
+        self.update_sun_position(utc_point_in_time)
+
+        # Set timer for the next solar event
+        async_track_point_in_utc_time(
+            self.hass, self.update_events,
+            self._next_change)
+        _LOGGER.debug("next time: %s", self._next_change)
 
     @callback
     def update_sun_position(self, utc_point_in_time):
         """Calculate the position of the sun."""
-        self.solar_azimuth = self.location.solar_azimuth(utc_point_in_time)
-        self.solar_elevation = self.location.solar_elevation(utc_point_in_time)
-
-    @callback
-    def point_in_time_listener(self, now):
-        """Run when the state of the sun has changed."""
-        self.update_sun_position(now)
-        self.update_as_of(now)
+        # Round azimuth to nearest 0.5
+        self.solar_azimuth = round(
+            self.location.solar_azimuth(utc_point_in_time), 2)
+        self.solar_elevation = round(
+            self.location.solar_elevation(utc_point_in_time), 2)
```

Comment:
Instead of using hard-coded update times, why not round the angles? Then the state machine would take care of de-duplicating values (because of force_update=False).

The sun position calculation is really quick - so computation time is not a problem. Database storage however is.

We could also have different rounding accuracy based on the value. So for example 

 - if abs(elevation) < 0.1: accuracy = 0.1
 - else: accuracy = 1.0

Reply:
I did try that. The thing is that the azimuth swings around rapidly at near noon, and so you end up with a bunch of pretty useless updates if you're tracking the azimuth too.

You do need to do that too at high latitudes, at the extreme case if you are at the north pole the elevation doesn't change much during 24hrs, only the azimuth. This affects what windows the dun shines into for example.

That's what lead to total arc of travel, rather than change with respect to the horizon. We do care more about changes near the horizon because they're more important, but mostly we want to know that the sun has moved somewhere new in the sky.

I'm not trying to minimise the calculation, the goal is to reduce the number of writes to the database.

Reply:
I've added a more detailed description above to what the code is doing. It does take a bit to get your head around :grin: